### PR TITLE
Updater singleflight

### PR DESCRIPTION
**Pull request by gabriel**
Updater singleflight

### CODE REVIEW

File: go/updater/updater.go [hunk 1]
```diff
@@ -310,8 +315,18 @@ func (u *Updater) apply(src string, dest string) (tmpPath string, err error) {
...
-	defer u.Unlock()
+	if u.running {
+		err = fmt.Errorf("Already running update")
+		u.Unlock()
+		return
+	}
+	u.running = true
+	u.Unlock()
+	defer func() {
+		u.running = false
```
Comment (maxtaco) [inline go/updater/updater.go]: I don't think it's race-safe to unset `u.running` without holding the lock.

Comment (maxtaco) [inline go/updater/updater.go]: I think this [singleflight class](https://github.com/golang/groupcache/blob/master/singleflight/singleflight.go) might be safer.

Comment (maxtaco) [inline go/updater/updater.go]: Hmm... but you want it to fail if you can't grab the lock.. Hmm... we might have to make a little lock-derived class that has a Lock and Unlock but Lock fails if already locked.

Comment (gabriel) [inline go/updater/updater.go]: ahh yes I was looking for that ... should have asked first

File: go/updater/updater.go [hunk 2]
```diff
@@ -310,8 +315,21 @@ func (u *Updater) apply(src string, dest string) (tmpPath string, err error) {
...
+		return
+	}
+	u.running = true
+	u.Unlock()
+
+	defer func() {
+		u.Lock()
+		u.running = false
+		u.Unlock()
+	}()
```
Comment (maxtaco) [inline go/updater/updater.go]: :+1:

File: go/updater/updater_test.go [hunk 3]
```diff
@@ -123,6 +124,37 @@ func TestUpdateCheckErrorIfLowerVersion(t *testing.T) {
...
+	}
+	wg.Wait()
+
+	_, errAfter := u.Update(nullUpdateUI{}, true, true)
+	if errAfter != nil {
+		t.Fatal(errAfter)
+	}
+
+	// We should have errored (already running) all but once
+	if errCount != times-1 {
```
Comment (maxtaco) [inline go/updater/updater_test.go]: How do you guarantee that attempts 2 through 5 will overlap with attempt 1?

**Pull request merged**